File: packages/protocol-parser/src/abiTypesToSchema.test.ts
import { describe, expect, it } from "vitest";
import { abiTypesToSchema } from "./abiTypesToSchema";

describe("hexToSchema", () => {
  it("converts hex to schema", () => {
    expect(abiTypesToSchema(["bool"])).toMatchInlineSnapshot(`
      {
        "dynamicFields": [],
        "staticFields": [
          "bool",
        ],
      }
    `);
    expect(abiTypesToSchema(["bool", "bool[]"])).toMatchInlineSnapshot(`
      {
        "dynamicFields": [
          "bool[]",
        ],
        "staticFields": [
          "bool",
        ],
      }
    `);
    expect(abiTypesToSchema(["bytes32", "int32", "uint256[]", "address[]", "bytes", "string"])).toMatchInlineSnapshot(`
      {
        "dynamicFields": [
          "uint256[]",
          "address[]",
          "bytes",
          "string",
        ],
        "staticFields": [
          "bytes32",
          "int32",
        ],
      }
    `);
  });
});

File: packages/services/pkg/protocol-parser/abiTypesToSchema_test.go
package protocolparser_test

import (
	protocolparser "latticexyz/mud/packages/services/pkg/protocol-parser"
	. "latticexyz/mud/packages/services/pkg/schema-type"
	"reflect"
	"testing"
)

func TestAbiTypesToSchema(testing *testing.T) {
	abiTypesList := [][]SchemaType{
		{
			BOOL,
		},
		{
			BOOL,
			BOOL_ARRAY,
		},
		{
			BYTES32,
			INT32,
			UINT256_ARRAY,
			ADDRESS_ARRAY,
			BYTES,
			STRING,
		},
	}

	expectedSchemas := []protocolparser.Schema{
		{
			StaticFields: []SchemaType{
				BOOL,
			},
			DynamicFields: []SchemaType{},
		},
		{
			StaticFields: []SchemaType{
				BOOL,
			},
			DynamicFields: []SchemaType{
				BOOL_ARRAY,
			},
		},
		{
			StaticFields: []SchemaType{
				BYTES32,
				INT32,
			},
			DynamicFields: []SchemaType{
				UINT256_ARRAY,
				ADDRESS_ARRAY,
				BYTES,
				STRING,
			},
		},
	}

	for i, abiTypes := range abiTypesList {
		schema := protocolparser.AbiTypesToSchema(abiTypes)
		if !reflect.DeepEqual(schema, expectedSchemas[i]) {
			testing.Errorf("expected schema to be %v, got %v", expectedSchemas[i], schema)
		}
	}
}
